ElectronClient/keymapHandler.ts

```diff
@@ -0,0 +1,130 @@
+	{ accelerator: 'CommandOrControl+G', command: 'goToAnything', label: 'Goto Anything', section: 'Tools' },
+	{ accelerator: 'F1', command: 'openWebsite', label: 'Website and documentation', section: 'Help' },
+];
+
+const initKeymap = () => {
+	const itemsByCommand: { [command: string]: keymapItem } = {};
+	for (let i = 0; i < defaultKeymap.length; i++) {
+		// We should clone the keymapItem before assigning it
+		// Otherwise we're going to mess up the defaultKeymap array
+		itemsByCommand[defaultKeymap[i].command] = { ...defaultKeymap[i] };
+	}
+
+	return itemsByCommand;
+};
+
+const getAccelerator = (command: string) => {
+	const keymapItem = keymap[command];
+
+	// Keymap entries without an Accelerator property allows users to add their own
+	// It's okay to provide a null Accelerator for menu items by default
+	return (keymapItem === undefined || keymapItem.accelerator === undefined) ? null : keymapItem.accelerator;
+};
+
+const setAccelerator = (command: string, accelerator: string) => {
+	keymap[command].accelerator = accelerator;
+};
+
+// TODO #2
+// Implement generateCustomKeymap function
+// Compares "keymap" and "defaultKeymap" objects and finds changes
+// Generates an object which can be stringified and saved as "keymap-desktop.json" file
+
+const overrideKeymap = (customKeymap: customKeymapItem[]) => {
+	return new Promise((resolve, reject) => {
+		for (let i = 0; i < customKeymap.length; i++) {
+			const item = customKeymap[i];
+
+			// Validation errors will be caught and logged to the console
+			if (!item.command) return reject(new Error(`Missing command on keymap item: ${JSON.stringify(item)}`));
+			if (!keymap.hasOwnProperty(item.command)) return reject(new Error(`Invalid command on keymap item: ${JSON.stringify(item)}`));
+			if (!item.accelerator) return reject(new Error(`Missing Accelerator on keymap item: ${JSON.stringify(item)}`));
+			if (!isAccelerator(item.accelerator)) return reject(new Error(`Invalid Accelerator on keymap item: ${JSON.stringify(item)}`));
+
+			// TODO #1
+			// Check if there are conflicting Accelerators
+
+			setAccelerator(item.command, item.accelerator);
+		}
+
+		resolve();
+	});
+};
+
+const isAccelerator = (accelerator: string) => {
+	let keyFound = false;
+	const parts = accelerator.split('+');
+
+	return parts.every((val, index) => {
+		const isKey = keyCodes.test(val);
+		const isModifier = modifiers.test(val);
+
+		if (isKey) {
+			// Key must be unique
+			if (keyFound) return false;
+			keyFound = true;
+		}
+
+		// Key is required
+		if (index === parts.length - 1 && !keyFound) return false;
+		return isKey || isModifier;
+	});
+};
+
+// TODO #3
+// Implement "resetKeymapItem(command)" function
+// Restores the default Accelerator for some command
+
+const resetKeymap = () => {
+	keymap = initKeymap();
+};
+
+let keymap = initKeymap();
+module.exports = { getAccelerator, setAccelerator, isAccelerator, overrideKeymap, resetKeymap };
```

Comment:
As the keymapHandler maintains some state I feel it would be best to define it has a class, to make it clear what's a property or not. Later you will want to inject a logger, maybe a dispatcher, etc. and for all this it's more convenient to have a class.

It's also a bit to be consistent with the rest of the application where services like this are defined as classes. See for example the services defined in lib/services. I think you should create something like this too and make it extend `BaseService`.

Reply:
Agreed! I'll refactor.